### DISCUSSION TITLE
Unable to perform nested queries in GraphQL

### DISCUSSION THREAD

**Discussion by seb-jones** [craftcms/cms#9029] [Ideas]
### Description

I'm fetching entries via GraphQL, and I am trying to filter them so that only entries with a certain field within a specified set of ranges are returned. Here is a simplified version of my GraphQL query, with the tail omitted for brevity:

```graphql
query ProductQuery($price: [QueryArgument] = null) {
  products: entries(price: $price) {
    ... on products_product_Entry {
      id
      title
```

 I am able to specify a single range by setting my GraphQL variables like so:

```json
{
    "price": ["and", ">= 0", "< 50"]
}
```

Which suggests to me that it works like the [Yii where clause operator format](https://www.yiiframework.com/doc/api/2.0/yii-db-queryinterface#where()-detail).

However, if I do the following to get any entries that lie within any of the given ranges:

```json
{
    "price": [
        "or",
        ["and", ">= 0", "< 50"],
        ["and", ">= 200", "< 250"]
    ]
}
```

An error is returned (trace omitted for brevity):

```json
"errors": [
    {
      "debugMessage": "QueryArgument must be either a string, an integer, or a boolean value.",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "trace": [
```

Is there a way to do complex queries such as this using GraphQL?

### Steps to reproduce

1. Set up an entry type with a numeric field and populate it with some entries.
2. Set up a GraphQL entry query and variables like the one above and run it.

### Additional info

- Craft version: Craft Pro 3.6.12
- PHP version: 7.3.27
- Database driver & version: MySQL 5.7.33
- Plugins & versions: Amazon S3 1.2.11, Feed Me 4.3.6, Seeder 3.3.1, SEOmatic 3.3.40

**Comment by brandonkelly** [comment]
> I am able to specify a single range by setting my GraphQL variables like so:
> 
> ```json
> {
>     "price": ["and", ">= 0", "< 50"]
> }
> ```
> 
> Which suggests to me that it works like the [Yii where clause operator format](https://www.yiiframework.com/doc/api/2.0/yii-db-queryinterface#where()-detail).

Unfortunately no; only a top-level `and` or `or` can be passed, and that’s where the similarities with Yii’s conditional syntax end.

We’d need to first add support for more advanced conditional syntax to element queries, before we could add it to GraphQL. I’ll mark this as an enhancement and we can look into the feasibility of it.

**Reply by ericdrosas87** [reply]
> only a top-level and or or can be passed

Is that to say that nested fields cannot be queried with the whole parent object returned? For example, if I have an entry-as-a-field on another entry type, can I write a custom resolver to match on a field in the nested entry-as-a-field and return the whole parent entry?

**Reply by brandonkelly** [reply]
@ericdrosas87 You can already do that via the `relatedToEntries` argument (see [Relations](https://craftcms.com/docs/3.x/graphql.html#relations)).

```graphql
query {
  entries(relatedToEntries: [{someField: "value"}]) {
    ...
  }
}
```

**Reply by ericdrosas87** [reply]
I think I understand what you're saying, but I can't seem to get it to work. For starters, I can't seem to properly associate the related entry. Looking at the documentation I assume in your example "someField" should be on the entry-as-a-field?

My use case:
- Let's say I have a parent entry type called "tour".
- "tour" has a field called "marker" of type matrix
- "marker" has a block type with a field called "cosmoObject" which is the wholly separate section/entry-as-a-field

This query works fine if I match on a field on "tour":
```
{
  entries(search: "*someSearchValue*") {
    ...tour
  }
}
```

If I search for a value that should be present on a field in the "cosmoObject" I get no results.

How would I craft this query in a way similar to your example?

I would have thought something like this would work:
```
{
  entries(relatedToEntries: [{section: "tours"}], search: "*someSearchValue*") {
    ...tour
  }
}
```

But I get no results. Thoughts?

**Reply by brandonkelly** [reply]
Sorry, yeah that’s not possible since the Tour entry isn’t directly related to the other `cosmoObject` entries.

Can you post your exact use case as a new Q&A discussion here, or on [Stack Exchange](https://craftcms.stackexchange.com/)?

**Reply by ericdrosas87** [reply]
@brandonkelly I actually posed the question here:

https://craftcms.stackexchange.com/questions/38012/how-do-i-write-a-graphql-query-to-search-through-sub-entries-entries-as-fields-a

Could I accomplish this with a custom resolver?

**Reply by andris-sevcenko** [reply]
@ericdrosas87 just responded to your post with some idea on how to accomplish this.

**Reply by ericdrosas87** [reply]
@andris-sevcenko thanks for your thorough answer! I am currently implementing a custom query based on your suggestions.

This next question is directed at anyone who can answer: What is the appropriate channel for reporting code example errors? I've noticed a few syntactical and other errors on the Extending GraphQL page.

**Comment by seb-jones** [comment]
Okay, thanks Brandon. Do you think there is any other way to achieve the intended functionality in GraphQL at the moment, or would it only be possibly in Twig/PHP?

**Comment by brandonkelly** [comment]
It’s only possible via [yii\db\Query::andWhere()](https://www.yiiframework.com/doc/api/2.0/yii-db-query#andWhere()-detail) currently.

Twig:

```twig
{% set products = craft.entries()
  .andWhere([
    'or',
    ['and', 'field_price >= 0', 'field_price < 50'],
    ['and', 'field_price >= 200', 'field_price < 250'],
  ])
  .all() %}
```

PHP:

```php
$products = \craft\elements\Entry::find()
    ->andWhere([
        'or',
        ['and', 'field_price >= 0', 'field_price < 50'],
        ['and', 'field_price >= 200', 'field_price < 250'],
    ])
    ->all();
```

**Comment by joshuabaker** [comment]
#7576

**Comment by FreekVR** [comment]
Just came here after running into similar issues, and having this type of advanced where clause in GQL would be really great. I created an entry type "Event" with a "Start" and "End" date field. I wanted to create a listing of events either starting after today OR ending after today, but I cannot currently add this kind of grouped OR condition to the queries.

For reference, this is how the ligthouse gql package for Laravel approaches this:
https://lighthouse-php.com/5/eloquent/complex-where-conditions.html#usage

I believe it uses the same underlying gql php library as Craft does, so perhaps this could be implemented in a similar way.

Implementing that in Craft would make the example @brandonkelly posted above look something like this in GraphQL:

```graphql
{
  entries(
    type: "products",
    where: {
      OR: [
        {
          AND: [
            { column: FIELD_PRICE, operator: GTE, value: "0" }
            { column: FIELD_PRICE, operator: LT, value: "50" }
          ]
        },
       {
          AND: [
            { column: FIELD_PRICE, operator: GTE, value: "200" }
            { column: FIELD_PRICE, operator: LT, value: "250" }
          ]
        }
      ]
    }
  ) {
    id, title
  }
}
```

**Comment by angelokeirsebilck** [comment]
Any updates on this? Is this something on the roadmap?

**Reply by brandonkelly** [reply]
I’m working on making it better for Craft 5, including for GraphQL ;)

**Comment by ericdrosas87** [comment]
Just an FYI I ended up writing[ this custom plugin ](https://plugins.craftcms.com/nested-entries-graphql-queries?craft3) for Craft v3.x and I think it would work with Craft 4 as well - though I haven't updated it as such.

**Comment by venciarn** [comment]
Hi, any update on this in Craft 5?
